hamt/shardeddir.go

```diff
@@ -0,0 +1,353 @@
+package hamt
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Stebalien/go-bitfield"
+	"github.com/ipfs/go-unixfsnode/data"
+	"github.com/ipfs/go-unixfsnode/iter"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+const (
+	// HashMurmur3 is the multiformats identifier for Murmur3
+	HashMurmur3 uint64 = 0x22
+)
+
+var _ ipld.Node = UnixFSHAMTShard(nil)
+var _ schema.TypedNode = UnixFSHAMTShard(nil)
+
+type UnixFSHAMTShard = *_UnixFSHAMTShard
+
+type _UnixFSHAMTShard struct {
+	ctx          context.Context
+	_substrate   dagpb.PBNode
+	data         data.UnixFSData
+	lsys         *ipld.LinkSystem
+	bitfield     bitfield.Bitfield
+	shardCache   map[ipld.Link]*_UnixFSHAMTShard
+	cachedLength int64
+}
+
+func NewUnixFSHAMTShard(ctx context.Context, substrate dagpb.PBNode, data data.UnixFSData, lsys *ipld.LinkSystem) (ipld.Node, error) {
+	if err := ValidateHAMTData(data); err != nil {
+		return nil, err
+	}
+	shardCache := make(map[ipld.Link]*_UnixFSHAMTShard, substrate.FieldLinks().Length())
+	bf := BitField(data)
+	return &_UnixFSHAMTShard{
+		ctx:          ctx,
```

Comment:
holding onto context past construction may be a mis-match that confuses things. currently context is passed in during a load and the caller may assume that's tracking the needed loads to the datastore to construct the node, such that once they get their constructed top-level node back they cancel the context.
if later they attempt to do a map iterator over the children in a directory they aren't necessarily going to make the connection that their previous construction-time context is still in play.

On the flip side, the whole point of the hamt is that we can lazy-load, and we do need a context of some sort while doing that, and I don't see any others to pull in that make any more sense.

thoughts @warpfork

Reply:
yes. the fix would be to pass context to all the ipld node methods, which is not a bad idea per say, but a real big change.

I mean the reality is all these loads are happening in LookupByString, which is a bit confusing I think, since who would expect that massive amounts of fetching from datastores and even the network is happening here. But, on the other hand, I don't have anything better right now.

Reply:
yup. understood that this is what the interface lets us do for now, but also pointing it out to @warpfork that the interface forces this in it's current design

Reply:
I would very much prefer that interfaces are changed to allow context. Storing in a struct temporarily is problematic in that once more code gets built on this the embedded context gets harder to remove, and the context's lifetime may be confusing.  If this cannot be changed now, please leave a TODO to indicate this is a workaround that needs to be addressed later.

Reply:
Interesting feedback, thanks for this.

- It is true that currently storing a context is really the only way it's possible to go with this.
- Agree that storing context means lifetime can be unclear and that seems kind of problematic.
- Confirm that adding contexts to all methods is a pretty big change.

@mvdan and I actually talked about this (the possibility of adding context parameters to everything) previously too, and at the time put it back down as insufficiently convinced that it provided value for the amount of line noise it adds.  (Mostly because the amount of syntactic bulk in these APIs already often feels like it's kind of pushing it.)  But it's definitely possible it's still worth reconsidering.

As an interesting comparison, the new golang `fs` packages also weighed a similar decision -- whether to add contexts to almost all file IO methods -- and they decided there to store context in a filesystem structure as well.  I don't know if that's right or wrong either, but it's food for thought.

Reply:
If it helps to think about things, this article describes why it is important to pass Context rather than store it in another type. It also highlights a rare case where storing Context in a struct type may make sense, and how to do so safely:
https://blog.golang.org/context-and-structs

Perhaps @mvdan could offer his take on this.  He has commented on this in the past: https://github.com/golang/go/issues/22602#issuecomment-430646394